### PR TITLE
Require matching hashlen in RSA functions: implementation

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
RSA and PK signature function that take both a `hash` parameter and a `hashlen` or `hash_len` parameter now require the length parameter to be the length of the input. If there is additional information about the hash algorithm and the length does not match, the functions return an error instead of silently ignoring the length parameter.

Continuation of https://github.com/ARMmbed/mbedtls/pull/4665 and https://github.com/ARMmbed/mbedtls/pull/4702, will need to be rebased when https://github.com/ARMmbed/mbedtls/pull/4665 is merged.

### CODE REVIEW

File: library/rsa.c
Comment (mpg) [inline library/rsa.c:2035]: Note: I'm not sure how that is related to the hashlen work. I this is an independent hardening, we should probably backport it?

File: tests/suites/test_suite_pk.function
Comment (mpg) [inline tests/suites/test_suite_pk.function:615]: Note: this is very similar to the second commit of https://github.com/ARMmbed/mbedtls/pull/4561 - so perhaps we can close #4561 in favour of this one ? (But probably still backport the relevant parts of #4561, see my comments there.)

**Review by mpg — APPROVED** [review]
This looks good to me.

I'm closing #4702 in favour of this PR in order to avoid diluting or review effort and having dependency chains, as this PR is no more risky than #4702 considering it passed CI.

I'm not closing #4665 now because it has an associated 2.x backport.

We should probably also close #4561 as the changes in this PR are a superset what it does, but someone should probably still backport the relevant parts of #4561 to 2.x and 2.16. In the interest of speed, I suggest you do it rather than the original contributor. If you agree, please close #4561 and let its author know what's going on.

File: library/rsa.c
Comment (gilles-peskine-arm) [reply library/rsa.c:2035]: Now that the functions have a clearly defined input length, I wanted to make the code robust to changes that could result in accepting different input lengths (for example, one day this code might run on a long SHAKE output). That meant checking that `hashlen` fits, regardless of any assumption we might make on hashes being small enough.

Indeed I think this change should be backported.

File: tests/suites/test_suite_pk.function
Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function:615]: I didn't look at #4561 while writing this, I'll take a look after #4665 is merged and I've rebased this.

Comment (yanesca) [reply tests/suites/test_suite_pk.function:615]: Since #4561 is a community PR and this PR (#4707) needs rebasing anyway, I think we could merge #4561 quickly and wait with rebasing #4707 until then. What do you think?

Comment (gilles-peskine-arm) [reply tests/suites/test_suite_pk.function:615]: #4561 is a behavior change where #4707 changes the behavior again, plus some testing improvements. I think it would be easier to put #4707 first and then make a version of #4561 with just the testing improvements, and perhaps some implementation improvements in `pk.c`, but no behavior change.

**Review by tom-daubney-arm — APPROVED** [review]
No issues. This all LGTM. Thanks.

File: library/rsa.c
Comment (gilles-peskine-arm) [reply library/rsa.c:2035]: The backports don't need to hold up the upcoming release. Filed as a Q3 tech debt item in https://github.com/ARMmbed/mbedtls/issues/4722

Comment (davidhorstmann-arm) [reply library/rsa.c:2035]: Hmm
Should this check be backported?
If `md_alg == MBEDTLS_MD_NONE` then we never reach it because we've already returned (in 2019), and if `md_alg != MBEDTLS_MD_NONE` then we have already performed this check in 1974, and none of the values have changed since.
Have I missed something here?

Comment (gilles-peskine-arm) [reply library/rsa.c:2035]: I (and the reviewers of this PR) missed that the check was already done earlier. So instead of backporting it, please make a PR on `development` to remove the redundant check!

Comment (mpg) [reply library/rsa.c:2035]: As a matter of fact, I noticed the check was already done earlier but noticed it was inside a conditional, and didn't go to the next step and failed to notice that this path was always taken if we ever get to the point where the new check was added. Should probably have said something, though. Thanks @davidhorstmann-arm for noticing.

**Pull request merged**
